File: SoojinKim/2023-10-08-Interval-List-Intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
        if(firstList.length==0 || secondList.length==0){
            return new int[0][0];
        } 
        int i = 0;
        int j = 0;
        int startMax = 0, endMin = 0;
        List<int[]> ans = new ArrayList<>();
        
        while(i<firstList.length && j<secondList.length){
            startMax = Math.max(firstList[i][0],secondList[j][0]);
            endMin = Math.min(firstList[i][1],secondList[j][1]);
            
			if(endMin>=startMax){           
                ans.add(new int[]{startMax,endMin});
            }
            
            if(endMin == firstList[i][1]) i++;       
            if(endMin == secondList[j][1]) j++;
        }
        
        return ans.toArray(new int[ans.size()][2]);
    }
}

File: MinguKim/2023-10-08-interval-list-intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        firstList.sort(key=lambda x: x[0])
        secondList.sort(key=lambda x: x[0])
        i,j = 0,0
        result=[]
        while i<len(firstList) and j<len(secondList):
            start = max(firstList[i][0], secondList[j][0])
            end = min(firstList[i][1], secondList[j][1])

            if start <= end:
                result.append([start,end])
                
            if firstList[i][1] < secondList[j][1]:
                i += 1
            else:
                j += 1
        return result

        